### ISSUE TITLE
New AWS regions like ap-east-1 cause aws.provider errors

### ISSUE THREAD

**Issue by rdanno** [pulumi/pulumi-aws#2194]
### What happened?

We have run into what seems like a showstopper of a problem, at least for us. [New regions are being provisioned](https://aws.amazon.com/premiumsupport/knowledge-center/iam-validate-access-credentials/) with session tokens v2 and this breaks our aws.Provider() with the following errors:
```
    Error: failed to refresh cached credentials, operation error STS: AssumeRole, failed to sign request: failed to retrieve credentials:
        raise invoke_error
    Exception: invoke of aws:index/getCallerIdentity:getCallerIdentity failed: invocation of aws:index/getCallerIdentity:getCallerIdentity returned an error: 1 error occurred:
    	* error configuring Terraform AWS Provider: no valid credential sources for Terraform AWS Provider found.
```
I know there is a newer aws_native library which may fix this, we’re not using that. (and resources don't seem to be cross compatible with the native provider)

### Steps to reproduce

Perform any call against a region endpoint using tokens v2 like ap-east-1 using a provider with a profile using sts:assumerole 

```
_provider = aws.Provider("Provider", region=region, profile="profile_name")
```

### Expected Behavior

It works

### Actual Behavior

```
    Error: failed to refresh cached credentials, operation error STS: AssumeRole, failed to sign request: failed to retrieve credentials:
        raise invoke_error
    Exception: invoke of aws:index/getCallerIdentity:getCallerIdentity failed: invocation of aws:index/getCallerIdentity:getCallerIdentity returned an error: 1 error occurred:
    	* error configuring Terraform AWS Provider: no valid credential sources for Terraform AWS Provider found.
```

### Additional context
Went thru botocore code, and this is where the SDK decides when to to use the regional or global STS endpoints:
https://github.com/boto/botocore/blob/dbc23f090d1257095da8bade8cb3fd5eeaec31db/botocore/args.py#L385-L388

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by rdanno** [comment]
This seems like it could be the same issue https://github.com/pulumi/pulumi-aws/issues/2188

**Comment by roothorp** [comment]
Hi @rdanno, thanks for the issue. I've raised this with the team to appropriately prioritize. You may be correct that it relates to #2188, so I'll make sure to discuss that too. Thanks!

**Comment by thomas11** [comment]
Hi @rdanno, could you tell us the versions of pulumi and the pulumi-aws provider you're using?

**Comment by rdanno** [comment]
```
CLI
Version      3.44.1
Go Version   go1.19.2
Go Compiler  gc

Plugins
NAME     VERSION
aws      5.19.0
command  0.5.2
python   unknown

Host
OS       darwin
Version  11.7
Arch     x86_64

This project is written in python: executable='/usr/local/bin/python3' version='3.9.13
'

Dependencies:
NAME            VERSION
ansible         6.5.0
pip             22.1.1
pulumi-aws      5.19.0
pulumi-command  0.5.2
pygount         1.4.0
setuptools      62.3.2
```

**Comment by thomas11** [comment]
Thanks @rdanno. I tried to reproduce this but wasn't able to, so far. I set up a basic Pulumi project with this config:

```
aws:assumeRole  {"roleArn":"arn:aws:iam::12345:role/user"}
aws:region      ap-east-1
```

First, it failed with

> error: unable to validate AWS credentials.
>    Details: IAM Role (arn:aws:iam::616138583583:role/tkappler) cannot be assumed.
>    [...]
>    Error: operation error STS: AssumeRole, https response error StatusCode: 403, RequestID: 00b54b6b-5225-4f08-9541-829758f0d1f7, api error AccessDenied: User: (redacted) is not authorized to perform: sts:AssumeRole on resource: (redacted)

I realized after a while that region ap-east-1 wasn't enabled for my account. I [enabled it in the aws console](https://aws.amazon.com/blogs/aws/now-open-aws-asia-pacific-hong-kong-region/) and that unblocked `pulumi up`.

Could you verify you have the region enabled and your pulumi config looks similar to the above?

**Comment by rdanno** [comment]
Hi thanks for looking into it!
I definitely have that region enabled.
Are you using a provider which calls a profile in your resource like this? 
`_provider = aws.Provider("Provider", region=region, profile="profile_name")`

**Comment by tiadobatima** [comment]
@thomas11: I'm wondering if this problem is only manifesting when using an assume-role profile, instead of direct pulumi configs?

In our case, with `ap-east-1` enabled, we get the error reported by @rdanno when using a profile in `~/.aws/config` similar to the one below and referencing the profile name in the pulumi code:

```
[profile prod]
role_arn = arn:aws:iam::123456789:role/some-role
source_profile = default
external_id = blablabla
role_session_name = my-session
```

As mentioned, the error happens in `ap-east-1` which uses a V2 STS endpoint, but it doesn't happen in regions where V1 is used (global STS endpoint).

The same error happens in much older versions of awscli/botocore. Newer versions of the awscli handle V2 regional STS endpoints transparently.
My guess is that the terraform provider used by pulumi is using an older version of the AWS SDK that doesn't handle the selection of regional endpoints transparently, as the newer versions of botocore do. But this is just guess 🤷‍♂️

**Comment by tiadobatima** [comment]
Thanks for trying that on your end, @thomas11 !

We're using python. Do you think that would make a difference?

**Comment by thomas11** [comment]
This works as well for me.

```python
import pulumi
import pulumi_aws as aws
from pulumi_aws import s3

provider = aws.Provider("provider", region="ap-east-1", profile="prod")

# Create an AWS resource (S3 Bucket)
bucket = s3.Bucket('my-bucket',
    opts=pulumi.ResourceOptions(provider=provider))

# Export the name of the bucket
pulumi.export('bucket_name', bucket.id)
```

I see `source_profile=default` in your config. What is default? Does it work if you that profile directly from your code?

**Comment by tiadobatima** [comment]
That default source_profile is the one we're keeping the access key for the IAM user assuming the role.
And yes, it does work if don't use an assume-role profile, and instead use a profile with a regular IAM user/access keys

**Comment by thomas11** [comment]
I can't think of much more I could do, having proven that ap-east-1 auth via `role_arn` works in general. You could try with a different AWS identity, or taking my repro sample and adapting it towards your scenario.

**Comment by aureq** [comment]
@thomas11 fwiw, maybe we should try on a brand new AWS account if we haven't already. Depending on the account age, services and APIs could be exposed/configured slightly differently.

**Comment by rdanno** [comment]
Hi @thomas11 thanks so much for looking into this.

In my [profile profile_name] i am using a external_id

I also noticed that you are using SSO rather than AWS keys. I wonder if that makes a difference in the authentication.

If it helps... i can reproduce this error in the latest version of aws-cli when hitting any STSv2 endpoint FIRST. (it fails to build a cache) If i try a regular region first, the cache is build, then the v2 region calls work after that. Seems like an aws-cli bug related to the handling of these new sts endpoints.

**Comment by rdanno** [comment]
After some more testing with aws-cli I can confirm that we cannot generate a session when starting with any v2 endpoint 

```An error occurred (InvalidClientTokenId) when calling the AssumeRole operation: The security token included in the request is invalid.```

If we go to an original v1 endpoint for any call then the session+cache is built; and this is what is used for subsequent authentication... which does not interact with STS again (until the session/cache expires)

I'll raise this with AWS. Certainly seems like the same thing is happening behind the scenes with this provider...

**Comment by thomas11** [comment]
That's a great find, @rdanno! Thanks for updating this issue.

**Comment by rdanno** [comment]
> That's a great find, @rdanno! Thanks for updating this issue.

No problem. Are you able to reproduce the aws-cli condition on your end? I'm confident if you see this happening with aws cli then you'll also see it happening with Pulumi in those regions. Otherwise it might be because you have a different auth method.

**Comment by rdanno** [comment]
Update: So we have found that enabling the same region on the _source_ account (the account with the user which is assuming the role) fixes this issue in aws-cli and pulumi and presumably everything else.

We noticed that assumerole calls are made to _both_ both accounts to the regional sts endpoint. It makes sense that if one of these is unavailable the process would fail. Contrary to what the documentation reads... we are following up with AWS for clarification. 

https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_enable-regions.html
The documentation there seems to indicate the user in account A does not need the regional endpoint enabled.
